### ISSUE TITLE
Tracking Issue for PathBuf::add_extension and Path::with_added_extension

### ISSUE THREAD

**Issue by tisonkun** [rust-lang/rust#127292]
Feature gate: `#![feature(path_add_extension)]`

This is a tracking issue for

* `PathBuf::add_extension`
* `Path::with_added_extension `
 
### Public API

```rust
impl PathBuf {
    pub fn add_extension<S: AsRef<OsStr>>(&mut self, extension: S) -> bool { ... )
}

impl Path {
    pub fn with_added_extension<S: AsRef<OsStr>>(&self, extension: S) -> PathBuf { ... }
}
```

### Steps / History

* [x]  ACP: https://github.com/rust-lang/libs-team/issues/368
* [x]  Implementation: https://github.com/rust-lang/rust/pull/123600/
* [ ]  Final comment period (FCP)
* [ ]  Stabilization PR

### Implementation history

- https://github.com/rust-lang/rust/pull/123600
- https://github.com/rust-lang/rust/pull/140163
 
### Unresolved Questions

None yet.

## Footnotes

1. https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html [↩](#user-content-fnref-1-39877e12402a53c830c980b62b917a0e)

**Comment by dilawar** [comment]
Hoping this becomes available in stable soon. Such a nice QoL api.

**Comment by anatawa12** [comment]
This new API is awesome!

But I think It would be better to rename (deprecate old and add new name) existing API `with_extension`  with `with_replaced_extension` and `set_extension` with `replace_extension` along with this feature.
I feel existing function `with_extension` and `set_extension` is confusing if it will replace or append extension.

Actually I mistakenly used `with_extension` as `with_added_extension` and we can see some such case in rust form like [this](https://users.rust-lang.org/t/append-an-additional-extension/23586/17).

**Comment by dvdsk** [comment]
> But I think It would be better to rename (deprecate old and add new name) existing API `with_extension` with `with_replaced_extension` and `set_extension` with `replace_extension` along with this feature.

I agree with the rename however we can only do that with a new edition (since `with_extension` is stable) right? And that might take a few more years. That should not block the stabilization of the `added` API.

**Comment by anatawa12** [comment]
I don't think edition change is required since we don't have to actually rename the function.
We can introduce `with_replaced_extension` in some version, and deprecate `with_extension` in following or later version.

I think there are some cases that deprecating without edition change like deprecate _ms functions with `Duration` parameter variants.

**Comment by joshtriplett** [comment]
Starting an FCP to stabilize this:

@rfcbot merge

**Comment by tgross35** [comment]
Some late bikeshed: the name `with_added_extension` feels very unnatural, what about the names `push_extension` and `join_extension`? This nicely mirrors what we have for path components:

```rust
impl Path {
    pub fn join<P: AsRef<Path>>(&self, path: P) -> PathBuf;
    pub fn join_extension<S: AsRef<OsStr>>(&self, extension: S) -> PathBuf { ... }
}

impl PathBuf {
    pub fn push<P: AsRef<Path>>(&mut self, path: P);
    pub fn push_extension<S: AsRef<OsStr>>(&mut self, extension: S) -> bool { ... )
}
```

**Comment by Amanieu** [comment]
We discussed this in the libs-api meeting. There was a slight preference towards keeping the "add"-based names since it is consistent with `set_extension` and `with_extension`.

**Comment by tbu-** [comment]
This API currently seems to have the same issue as `Path::set_extension` had before introducing a panic in #125070.

This API should probably panic as well if the passed extension contains a path separator.

**Comment by tbu-** [comment]
Additional, but much less important point: The `add_extension` API returns a boolean which is likely not checked at call site to communicate its success in adding the extension. I think it'd be a lot better if it returned a `Result` which would make the user think about what to do if the file extension could not be set.

This is exemplified in the `with_added_extension` API which promptly forgets to check the result of `add_extension` and currently also has no way to return that result to the user.

**Comment by T0mstone** [comment]
> I think it'd be a lot better if it returned a `Result` which would make the user think about what to do if the file extension could not be set.

Alternatively, the function could just be annotated with `#[must_use]`.

**Comment by tisonkun** [comment]
> This API currently seems to have the same issue as Path::set_extension had before introducing a panic in https://github.com/rust-lang/rust/pull/125070.

This sounds reasonable to keep the manner in sync. As you're the author of #125070, would you submit a PR to apply the same for `add_extension`? Otherwise, I can submit one.

**Comment by tbu-** [comment]
> * It should return a `bool` just like `set_extension`. However we opted not to make this `#[must_use]` because the return value seems little-used: Github code search shows ~18k uses of `set_extension`, **none of which check the return value.**

This seems to indicate a problem. Apparently it's common to disregard the error. In fact, with `with_extension` and `with_added_extension`, there's no way to check for the error. Why is that okay? Do the 18k users really want to explicitly ignore the error?

**Comment by dtolnay** [comment]
In my experience using `set_extension`, they do want to ignore, because the failure case is unreachable. For example: crawling a directory (or glob) to look for `*.rs` files, and changing the extension to `.rs.h`. In this scenario it's impossible for one of the filenames you are setting the extension of to be `/` or `..` or `.`.

**Comment by Timmmm** [comment]
I agree with @tbu-  and @Amanieu:

1. `with_extension()` should be renamed. When I first read the name I thought it would add an extension, not replace it. I think `with_replaced_extension()` or `with_extension_replaced()` or `with_new_extension()` are all better.
2. `with_replaced_extension()` should probably return `Result<(), ...>`. Most people will just throw an `.unwrap()` or `.expect()` or `?` on there, but I definitely didn't think about the possibility that it could fail until I read this discussion. And yeah, it's unreachable but I want to *opt in* to that.

**Comment by thaliaarchi** [comment]
> We discussed this in the libs-api meeting today. The consensus was to keep this API as close to the existing `set_extension` and `with_extension` as possible. This means that: It should panic if the extension contains a path separator.

I've implemented that in https://github.com/rust-lang/rust/pull/140163.

**Comment by hanna-kruppe** [comment]
Stabilization PR: #145209

**Comment by hanna-kruppe** [comment]
> I don't think anyone has disagreed that:

People (including libs-api members) very much have disagreed, several times, in the above discussion and in meetings reported in comments on this issue. Just not in response to your comment specifically because very similar points were raised by others before. That doesn’t mean the team can’t change its mind now, but it’s not new information either.
